core/src/main/java/org/kohsuke/stapler/RequestImpl.java

```diff
@@ -800,68 +814,80 @@ private Object bindResolve(Object o, JSONObject src) {
-                        }
-                    }
+        final Set<String> keys = j.keySet();
+
+        // try field injection first
+        for (Class c=r.getClass(); c!=null; c=c.getSuperclass()) {
+            Arrays.stream(c.getDeclaredFields())
+                .filter(f -> f.isAnnotationPresent(DataBoundSetter.class))
+                .forEach(f -> {
+                    final String name = f.getName();
+                    if (exclusions.contains(name)) return;
+                    setField(r, j, name, f);
+                });
+        }
 
-                    Method wm = findDataBoundSetter(r.getClass(), key);
-                    if (wm==null)   continue;
+        for (Class c = r.getClass() ; c!=null; c=c.getSuperclass()) {
+            Arrays.stream(c.getMethods())
+                .filter(m -> m.getParameterCount() == 1)
+                .filter(m -> m.getName().startsWith("set"))
+                .filter(m -> m.isAnnotationPresent(DataBoundSetter.class))
+                .forEach(m -> {
+                    String propertyName = Introspector.decapitalize(m.getName().substring(3));
+                    if (exclusions.contains(propertyName)) return;
+                    invokeSetter(r, j, propertyName, m);
+                });
+        }
 
-                    Class<?>[] pt = wm.getParameterTypes();
-                    assert pt.length==1;
 
-                    // only invoking public methods for security reasons
-                    wm.invoke(r, bindJSON(wm.getGenericParameterTypes()[0], pt[0], j.get(key)));
-                } catch (IllegalAccessException e) {
-                    LOGGER.log(WARNING, "Cannot access property " + key + " of " + r.getClass(), e);
-                } catch (InvocationTargetException e) {
-                    LOGGER.log(WARNING, "Cannot access property " + key + " of " + r.getClass(), e);
-                }
-            }
-        }
+       final Set violations = validator.validate(r);
+        if (!violations.isEmpty()) throw new ConstraintsValidationException(violations);
 
         invokePostConstruct(getWebApp().getMetaClass(r).getPostConstructMethods(), r);
 
         return r;
     }
 
-    private Method findDataBoundSetter(Class c, String name) {
-        // look for public setter that has the matching name
-        for ( ; c!=null; c=c.getSuperclass()) {
-            for (Method m : c.getDeclaredMethods()) {
-                if (!Modifier.isPublic(m.getModifiers())
-                 || !m.getName().startsWith("set")
-                 || m.getParameterTypes().length!=1
-                 || !m.isAnnotationPresent(DataBoundSetter.class))
-                    continue;
+    private <T> void setField(T r, JSONObject j, String name, Field f) {
+        final Object json = getJsonValueOrDefault(j, name, f);
+        if (json == null) return;
+        final Object value = bindJSON(f.getGenericType(), f.getType(), json);
+        try {
+            f.setAccessible(true);
```

Comment:
Didn't you yourself publish a library to avoid having to call this directly? Could be useful here.

Reply:
I indeed experimented with java 9's `VarHandle` but this is not relevant here, `setAccessible` is well supported in java 9+, just reflexion on _internal_ JRE classes is forbidden.
I also try to keep this minimal, to ensure approving this PR won't drive into infinite debates 😇

Reply:
Ah ok, sounds good!